### PR TITLE
Enable Shiro in the fcrepo-webapp via Spring configuration

### PR DESCRIPTION
**Pull request by mohideen**
Resolves: https://jira.duraspace.org/browse/FCREPO-2763

# What does this Pull Request do?
Enable Shiro in the fcrepo-webapp via Spring configuration.

# Additional Notes:
Some the IT tests had to be Ignored until FCREPO-2760 is resolved.
Fixed AbstractPrincipalProvider bug that was revealed while running the maven tests with spring configuration.
Changes:
- The principals returned from the asSet() is unmodifiable. Creating a new HashSet from the returned set.
- Check for the null condition on Principal object retrieved from the session.
- Update/Set the principals to session only when there are new Principals.

# Interested parties
@fcrepo4/committers, @peichman-umd

### CODE REVIEW

File: fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml [hunk 1]
```diff
@@ -96,6 +96,60 @@
...
+
+    <!-- Post processor that automatically invokes init() and destroy() methods -->
+    <bean id="lifecycleBeanPostProcessor" class="org.apache.shiro.spring.LifecycleBeanPostProcessor"/>
+
+    <!-- Authentication Filter -->
+    <bean id="servletContainerAuthFilter" class="org.fcrepo.auth.common.ServletContainerAuthFilter"/>
+
+    <!-- Principal Provider Filters -->
+    <bean id="httpHeaderPrincipalProvider" class="org.fcrepo.auth.common.HttpHeaderPrincipalProvider"/>
+    <bean id="delegateHeaderPrincipalProvider" class="org.fcrepo.auth.common.DelegateHeaderPrincipalProvider"/>
```
Comment (whikloj) [inline fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml]: Already defined above on line 74

File: fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml [hunk 2]
```diff
@@ -96,6 +96,60 @@
...
+    </bean>
+
+    <!-- Post processor that automatically invokes init() and destroy() methods -->
+    <bean id="lifecycleBeanPostProcessor" class="org.apache.shiro.spring.LifecycleBeanPostProcessor"/>
+
+    <!-- Authentication Filter -->
+    <bean id="servletContainerAuthFilter" class="org.fcrepo.auth.common.ServletContainerAuthFilter"/>
+
+    <!-- Principal Provider Filters -->
+    <bean id="httpHeaderPrincipalProvider" class="org.fcrepo.auth.common.HttpHeaderPrincipalProvider"/>
```
Comment (whikloj) [inline fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml]: Already defined (but commented out) line 52

File: fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml [hunk 3]
```diff
@@ -96,6 +96,60 @@
...
+    <bean id="delegateHeaderPrincipalProvider" class="org.fcrepo.auth.common.DelegateHeaderPrincipalProvider"/>
+    <bean id="containerRolesPrincipalProvider" class="org.fcrepo.auth.common.ContainerRolesPrincipalProvider"/>
```
Comment (whikloj) [inline fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml]: Already defined (but commented out) line 60

**Comment by whikloj** [comment]
@peichman-umd you have a better grasp of this WebAC configuration, do you know what of the beans defined previously we will still need? Or should we just make a ticket to clean it out once Shiro is in place?

**Comment by whikloj** [comment]
This PR also seems to put the repository in an unusable state. I still get prompted to authenticate, but once I do all I get is a white screen. No matter the request.
```
[whikloj@net227.lib.umanitoba.ca 10:36 ] 
[/Users/whikloj] 
> curl -I http://localhost:8080/rest
HTTP/1.1 401 Unauthorized
Date: Wed, 25 Apr 2018 15:36:20 GMT
WWW-Authenticate: basic realm="fcrepo"
Content-Type: text/html;charset=iso-8859-1
Cache-Control: must-revalidate,no-cache,no-store
Content-Length: 326
Server: Jetty(9.3.1.v20150714)

[whikloj@net227.lib.umanitoba.ca 10:36 ] 
[/Users/whikloj] 
> curl -I http://localhost:8080/rest -ufedoraAdmin:fedoraAdmin
HTTP/1.1 200 OK
Date: Wed, 25 Apr 2018 15:36:26 GMT
Set-Cookie: JSESSIONID=13yow8woj94lb1mdnkeudqrk9i;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Tue, 24-Apr-2018 15:36:26 GMT
Content-Length: 0
Server: Jetty(9.3.1.v20150714)

[whikloj@net227.lib.umanitoba.ca 10:36 ] 
[/Users/whikloj] 
> curl -I http://localhost:8080/rest -ufedoraAdmin:fedoraAdmin
HTTP/1.1 200 OK
Date: Wed, 25 Apr 2018 15:36:37 GMT
Set-Cookie: JSESSIONID=1okxth8vnv2ma7v13l4jhvz2j;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Tue, 24-Apr-2018 15:36:37 GMT
Content-Length: 0
Server: Jetty(9.3.1.v20150714)

[whikloj@net227.lib.umanitoba.ca 10:36 ] 
[/Users/whikloj] 
> curl -I http://localhost:8080/rest/testobject -ufedoraAdmin:fedoraAdmin
HTTP/1.1 200 OK
Date: Wed, 25 Apr 2018 15:36:58 GMT
Set-Cookie: JSESSIONID=12lmhjb8wqlyv6r4g0fye4i2k;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Tue, 24-Apr-2018 15:36:58 GMT
Content-Length: 0
Server: Jetty(9.3.1.v20150714)

[whikloj@net227.lib.umanitoba.ca 10:36 ] 
[/Users/whikloj] 
> curl -I http://localhost:8080/rest/happygolucky -ufedoraAdmin:fedoraAdmin
HTTP/1.1 200 OK
Date: Wed, 25 Apr 2018 15:37:09 GMT
Set-Cookie: JSESSIONID=25nqrnley17kd3zsri2z7944;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Tue, 24-Apr-2018 15:37:09 GMT
Content-Length: 0
Server: Jetty(9.3.1.v20150714)
```

For reference I know `happygolucky` is not a valid URI in the fcrepo data.

**Comment by peichman-umd** [comment]
@whikloj I believe the only WebAC bean we will need in the repo.xml is the WebACRolesProvider, since I am still using that to do the actual authz logic in the WebACAuthorizingRealm. The other WebAC-related beans in repo.xml should be able to be removed.

**Comment by mohideen** [comment]
@whikloj I think it is best to wait until FCREPO-2760 is completed. Once that is in, I will merge those changes and test the repository and update the PR accordingly.

**Comment by whikloj** [comment]
@mohideen ok, I'll leave this until https://jira.duraspace.org/browse/FCREPO-2760 is in place.

**Comment by mohideen** [comment]
Created a new PR after rebasing with changes from FCREPO-2760.

See #1348

**Pull request closed without merging**